lib/spack/spack/hooks/__init__.py

```diff
@@ -2,78 +2,88 @@
-
 """This package contains modules with hooks for various stages in the
-   Spack install process.  You can add modules here and they'll be
-   executed by package at various times during the package lifecycle.
-
-   Each hook is just a function that takes a package as a parameter.
-   Hooks are not executed in any particular order.
-
-   Currently the following hooks are supported:
-
-      * pre_install(spec)
-      * post_install(spec)
-      * pre_uninstall(spec)
-      * post_uninstall(spec)
-      * on_install_failure(exception)
-
-   This can be used to implement support for things like module
-   systems (e.g. modules, lmod, etc.) or to add other custom
-   features.
+Spack install process.  You can add modules here and they'll be
+executed by package at various times during the package lifecycle.
+
+Each hook is just a function that takes a package as a parameter.
+Hooks are not executed in any particular order.
+
+Currently the following hooks are supported:
+
+    * pre_install(spec)
+    * post_install(spec)
+    * pre_uninstall(spec)
+    * post_uninstall(spec)
+    * on_install_start(spec)
+    * on_install_success(spec)
+    * on_install_failure(spec)
+    * on_phase_success(pkg, phase_name, log_file)
+    * on_phase_error(pkg, phase_name, log_file)
+    * on_phase_error(pkg, phase_name, log_file)
+    * on_analyzer_save(pkg, result)
+
+This can be used to implement support for things like module
+systems (e.g. modules, lmod, etc.) or to add other custom
+features.
 """
-import os.path
-
+import llnl.util.lang
 import spack.paths
-import spack.util.imp as simp
-from llnl.util.lang import memoized, list_modules
 
 
-@memoized
-def all_hook_modules():
-    modules = []
-    for name in list_modules(spack.paths.hooks_path):
-        mod_name = __name__ + '.' + name
-        path = os.path.join(spack.paths.hooks_path, name) + ".py"
-        mod = simp.load_source(mod_name, path)
-
-        if name == 'write_install_manifest':
-            last_mod = mod
-        else:
-            modules.append(mod)
-
-    # put `write_install_manifest` as the last hook to run
-    modules.append(last_mod)
-    return modules
-
-
-class HookRunner(object):
+class _HookRunner(object):
+    #: Stores all hooks on first call, shared among
+    #: all HookRunner objects
+    _hooks = None
 
     def __init__(self, hook_name):
         self.hook_name = hook_name
 
+    @classmethod
+    def _populate_hooks(cls):
+        # Lazily populate the list of hooks
+        cls._hooks = []
+        relative_names = list(llnl.util.lang.list_modules(
+            spack.paths.hooks_path
+        ))
+
+        # We want this hook to be the last registered
+        relative_names.sort(key=lambda x: x == 'write_install_manifest')
+        assert relative_names[-1] == 'write_install_manifest'
+
+        for name in relative_names:
+            module_name = __name__ + '.' + name
+            module_obj = __import__(module_name, fromlist=[None])
```

Comment:
Two things:

* https://stackoverflow.com/questions/2724260/why-does-pythons-import-require-fromlist and https://bugs.python.org/issue9254 mention `When importing a module from a package, note that __import__('A.B', ...) returns package A when fromlist is empty, but its submodule B when fromlist is not empty.` I think that's worth commenting on here to help future readers. FWIW the latter link has some discussion that when a user is tempted to use `__import__` with "junk" that they should instead use `importlib.import_module`
* I think this is the primary qualitative change: using `__import__` instead of `simp.load_source(mod_name, path)`. If I understand correctly, replacing the `memoize` with a class-level cache is just for readability - or am I wrong and that is also required?

Reply:
> FWIW the latter link has some discussion that when a user is tempted to use __import__ with "junk" that they should instead use importlib.import_module

I would totally agree with that, since to my understanding it will give us a better API (there's no need for the `fromlist=[None]` trick). Unfortunately we have to support Python 2.6 hence this solution...

Reply:
> I think this is the primary qualitative change: using __import__ instead of simp.load_source(mod_name, path). If I understand correctly, replacing the memoize with a class-level cache is just for readability - or am I wrong and that is also required?

That's correct.